Support model evaluation with an endpoint URL

B-Step62 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/B-Step62/mlflow/pull/11262?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11262/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11262
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

This change enhances [mlflow.evaluate()](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate) function to accepts an endpoint URL to `model` argument. We start to get frequent requests for this from users who want to evaluate the proprietary model endpoint or Databricks served models. Without this feature, they need to implement a custom predict function, which is not super difficult but less convenient.

**Important notes**
- We only support the OpenAI-compatible chat/completion models, to avoid asking too many configurations to users. 
- The input dataset should contain a single text input and MLflow constructs chat/completion request from it. This limits some flexibility for input e.g. system prompt for chat model, but is consistent with other flavors like OpenAI and Transformers. The more complex use case should be handled by [a custom function](https://mlflow.org/docs/latest/llms/llm-evaluate/index.html#evaluating-with-a-custom-function). 
- Inference params is not supported in current API, but they are indispensable for chat/completion models. This PR adds a new argument `inference_params` to the API, but it is only used for endpoint evaluation, not for other type of models. We can do follow-up to support inference params for all model types.

**Example**
```
eval_data = pd.DataFrame(
    {
        # Input data must be a string column and named "inputs".
        "inputs": [
            "Write 3 reasons why you should use MLflow?",
            "Can you explain the difference between classification and regression?",
        ],
    }
)


with mlflow.start_run() as run:
    results = mlflow.evaluate(
        # URL of the OpenAI-compatible chat/completion endpoint
        model="https://example.com/serving-endpoints/databricks-mpt-30b-instruct/invocations",
        data=eval_data,
        # New argument for support inference params
        inference_parameters={"max_tokens": 100, "temperature": 0.0},
        # New argument for passing auth info
        headers={"Authorization": "Bearer YOUR API KEY"},
        # By specifying the "completion" model type, MOLflow automatically format the input text to the completion request format,
        # and extract the generated text from the completion response.
        model_type="completion",
    )
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (WIP)

Tested in Databricks. Sample code for one test case:
```
import mlflow
import pandas as pd

eval_data = pd.DataFrame(
    {
        "inputs": [
            "What is MLflow?",
            "What is Spark?",
        ],
        "ground_truth": [
            "MLflow is an open-source platform ....",
            "Apache Spark is an open-source, ...",
        ],
    }
)


with mlflow.start_run(run_name="mlflow-eval-chat") as run:
    results = mlflow.evaluate(
        # Pass the URL of an OpenAI-compatible chat endpoint
        model="https://${{DB_HOST}}/serving-endpoints/databricks-mixtral-8x7b-instruct/invocations",
        data=eval_data,
        targets="ground_truth",
        model_type="chat",
        inference_params={"max_tokens": 100, "temperature": 0.0},
        headers={"Authentication": "Bearer ${{DB_TOKEN}}"},
        extra_metrics=[mlflow.metrics.exact_match()],
    )
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [x] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

The `mlflow.evaluate()` API supports a model endpoint URL as a evaluation subject. This is particularly useful when you want to evaluate deployed models or proprietary APIs such as [Databricks Foundation Model APIs](https://docs.databricks.com/en/machine-learning/foundation-models/index.html).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

B-Step62 committed:
Support Deployment Server endpoint for model evaluation

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
use model_type

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
docs

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
small fix

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,104 @@ up OpenAI authentication to run the code below.
 
+.. _llm-eval-model-endpoint:
+
+Evaluating with a Model Endpoint URL
```
We should use an `endpoints:/` URI, not a URL

B-Step62 replied:
Yea I found many utilities are implemented there and using the endpoints URI simplifies the logic a lot. Thanks for the clarification!

harupy commented on mlflow/models/evaluation/llm_utils.py:
```diff
@@ -0,0 +1,132 @@
+    "Invalid response format:\n{response}\n MLflow Evaluation expects OpenAI-compatible "
+    "response format for chat and completion models. Please refer to OpenAI documentation: "
+    "https://platform.openai.com/docs/api-reference/introduction for the expected format.\n\n"
+    "HINT: If your endpoint has a different response format, you can still use  it for evaluation "
```
```suggestion
    "HINT: If your endpoint has a different response format, you can still use it for evaluation "
```

harupy commented on mlflow/models/evaluation/llm_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+
+class _ServedLLMEndpointModel(mlflow.pyfunc.PythonModel):
+    """ """
```
```suggestion
```

harupy commented on tests/evaluate/test_evaluation.py:
```diff
@@ -1567,3 +1567,163 @@ def test_evaluate_with_static_dataset_error_handling_pandas_dataset():
+}
+
+
+@mock.patch("mlflow.models.evaluation.llm_utils.requests.post")
```
Can we use `with mock.patch` instead to minize the scope of monkey-patching?


```python
def test_foo():
    a()
    with mock.patch(...):
        b()
    c()
```

is better than:

```
@mock.patch(...)
def test_foo():
    a()
    b()
    c()
```

B-Step62 replied:
Hmm but annotation also has pros of making the code flat and telling what modules are mocked for the test case? Also it helps catch unexpected call to the module.

harupy replied:
@B-Step62 

> making the code flat (and prep -> run -> validate structure)

Can we achieve the same thing using `with mock.patch()`?

```
with mock.patch("foo.bar", return_value=foo) as mock_foo:  # prep
    do_foo()  # run

    mock_foo.xxx_yyy()  # validate
```

I'm not a big fan of the decorator `mock.patch` because I need to guess which lines require mocks and want to minimize the scope of monkey-patching to avoid false positives.

B-Step62 replied:
Yea I'm fine with both, just wanted to avoid nesting `with` statement like this. But it's not too ugly in this case after all.

```
with mock.patch(...) as mock_deploy_client:
     # do some mock setup
     with mlflow.start_run():
        ...
```

Speaking about the annotator itself, my position is that there is no clear winner between two options (e.g. false-positive can happen in the other way - let's say we want to mock and count calls to tracking server, then it may call to local server outside the content). The annotator works better for structuring the test method evenly, particularly if you have many test cases with multiple mocks, it much easier to read. But specifically to this case I don't see much difference anyway.

harupy replied:
@B-Step62 Fair enough. Let's use the decorator :)

harupy commented on mlflow/models/evaluation/llm_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+def get_model_from_llm_endpoint_url(
+    endpoint: str,
+    model_type: _ModelType,
```
```suggestion
    model_type: str,
```

harupy replied:
or 

```suggestion
    model_type: Literal[...],
```

harupy commented on mlflow/models/evaluation/llm_utils.py:
```diff
@@ -0,0 +1,132 @@
+            **params,
+        }
+
+    def extract_output(self, response) -> str:
+        try:
+            text = response["choices"][0]["message"]["content"]
+        except (KeyError, TypeError) as e:
+            raise MlflowException(HINT_MSG.format(response=response)) from e
+        return text
```
```suggestion
    def extract_output(self, response) -> str:
        try:
            return response["choices"][0]["message"]["content"]
        except (KeyError, TypeError) as e:
            raise MlflowException(HINT_MSG.format(response=response)) from e
```

B-Step62 committed:
use deployment client

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
fix

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
typo

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -403,14 +403,15 @@ Prepare Your LLM for Evaluating
    * Has ``data`` as the only argument, which can be a ``pandas.Dataframe``, ``numpy.ndarray``, python list, dictionary or scipy matrix.
    * Returns one of ``pandas.DataFrame``, ``pandas.Series``, ``numpy.ndarray`` or list. 
-3. Set ``model=None``, and put model outputs in `data`. Only applicable when the data is a Pandas dataframe.
+3. An MLflow Deployment endpoint URI pointing to a local `MLflow Deployment Server <../deployments/index.html>`_ or `Databricks Foundation Models <https://docs.databricks.com/en/machine-learning/foundation-models/index.html>`_. 
```
```suggestion
3. An MLflow Deployments endpoint URI pointing to a local `MLflow Deployment Server <../deployments/index.html>`_ or `Databricks Foundation Models <https://docs.databricks.com/en/machine-learning/foundation-models/index.html>`_. 
```

dbczumar replied:
@B-Step62 MLflow Deployments also supports Databricks External Models. Can we add that?

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,91 @@ up OpenAI authentication to run the code below.
 
+.. _llm-eval-model-endpoint:
+
+Evaluating with a MLflow Deployment Endpoint
```
```suggestion
Evaluating with an MLflow Deployments Endpoint
```

dbczumar replied:
Can we pluralize MLflow Deployments everywhere?

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,91 @@ up OpenAI authentication to run the code below.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For MLflow >= 2.11.0, :py:func:`mlflow.evaluate()` supports evaluating a model endpoint by directly passing the MLflow Deployment endpoint URI to the ``model`` argument.
+This is particularly useful when you want to evaluate a deployed model hosted by a local `MLflow Deployment Server <../deployments/index.html>`_ or `Databricks Foundation Model APIs <https://docs.databricks.com/en/machine-learning/model-serving/score-foundation-models.html#language-MLflow%C2%A0Deployments%C2%A0SDK>`_, without implementing custom prediction logic to wrap it as an MLflow model or a ...
```
Can we mention external models everywhere that we mention foundation models?

B-Step62 replied:
Sure, I thought the term "Foundation Model" includes it haha

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -1769,6 +1813,16 @@ def pred_sample(eval_df, _builtin_metrics, _artifacts_dir):
     from mlflow.utils import env_manager as _EnvManager
 
+    # Inference params are currently only supported for passing a deployment endpoint as the model.
+    # TODO: We should support inference_params for other model types
+    is_model_endpoint_url = isinstance(model, str) and model.startswith("endpoints:/")
```
I think `startswith` is somewhat brittle, e.g. 3 slashes is technically also valid. Can we use urlparse and check the scheme?

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -1769,6 +1813,16 @@ def pred_sample(eval_df, _builtin_metrics, _artifacts_dir):
+    if not is_model_endpoint_url and inference_params is not None:
+        raise MlflowException(
+            message="The inference_params argument can only be specified when the model "
+            "is an MLflow Deployments endpoint URI like `endpoints:/xxx`",
```
Nittiest of nits: can we use an example endpoint name rather than `xxx`?

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -1247,6 +1247,44 @@ def predict(self, context, model_input: pd.DataFrame):
+                )
+
+            predictions = []
+            for input_data in model_input["inputs"]:
+                if not isinstance(input_data, str):
```
Why couldn't the input data be a list of dictionaries corresponding to the OpenAI "messages" input? Seems necessary to support this if we want users to be able to evaluate an endpoint on a dataset of message histories, which is common for chatbots.

B-Step62 replied:
The main intention here is to start with simple interface and also consistent with OpenAI model flavors. But I agree we can support here (with clear documentation).

dbczumar replied:
Got it. Yeah, we should support message histories and include an example :)

B-Step62 replied:
Dictionary inputs needs some patching for dataset digest computation logic. I think it's not too dirty so ok, but otherwise we might want to leave this as follow-up.

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -1247,6 +1247,44 @@ def predict(self, context, model_input: pd.DataFrame):
+            self.params = params
+
+        def predict(self, context, model_input: pd.DataFrame):
+            if "inputs" not in model_input.columns:
```
What happens if the user passes `data` to `mlflow.evaluate()` of type list and then uses (or does not use) the `feature_names` argument? Does this fail? We should definitely support passing a list & test that case as well

B-Step62 replied:
I agree that we should support passing list input, but the usage of `feature_names` is still not clear to me after reading the docs tbh.... let me clarify my understanding from reading codes:
* When `feature_names` is provided for DataFrame input, those columns are picked to prediction input.
* When `feature_names` is NOT provided for DataFrame input, all columns except `target` will be picked to input.
* When `feature_names` is provided for list input, the list will be converted to dataframe with those given column names.
* When `feature_names` is NOT provided for list input, the list will be input with automatically generated column names `feature_names_0`, ...

Is my understanding correct? If my understanding is correct, supporting them is not super hard, with some limitation (like `feature_names` should only contain single input column).

B-Step62 replied:
Added support for list input in the latest revision. One friction I found is that it doesn't accepts the flat list like `["What is MLflow?", "What is Spark"]`, instead each element has to be wrapped in another list like `[["What is MLflow?"], ["What is Spark?"]]`. This is not specific to endpoint evaluation.

harupy commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,91 @@ up OpenAI authentication to run the code below.
+
+.. code-block:: python
+
+    from mlflow.deployments import set_deployments_target
```
Can we also import `pandas` and `mlflow` so users can copy & paste & execute this code quickly?

B-Step62 replied:
great idea!

B-Step62 committed:
Feedbacks:
1. Add support for list input
2. Fix issue of None targets handling to unblock 1
3. Add support for a dictionary element in input data, to directly pass chat/completion payload
4. Fix issue of dictionary hashing to unblock 2
5. Other comments

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+                ["What is Spark?"],
+            ]
+
+      - The :py:func:`mlflow.evaluate()` will also accepts a list input. One notable requirement is that the each
```
note: This is not new limitation but the current behavior of evaluate API. It's possible to support flat list, but not sure if it breaks other existing use cases, so just keep the behavior in this PR.

dbczumar replied:
Can we add that flat list support?
Hopefully existing CI will catch issues?

B-Step62 commented on mlflow/models/evaluation/base.py:
```diff
@@ -425,9 +425,42 @@ def _hash_uint64_ndarray_as_bytes(array):
+    return False
+
+
+def _hash_array_of_dict_as_bytes(data):
```
note: This hashing support is required for passing a list of dictionary (chat/completion payload) or similar pandas DataFrame as a input dataset. The hashing is required for generating dataset digest, but it doesn't handle a list or array contains dictionary. The implementation is not super clean, just band-aiding without breaking existing use cases.

B-Step62 commented on mlflow/models/evaluation/base.py:
```diff
@@ -1140,7 +1173,9 @@ def _convert_data_to_mlflow_dataset(data, targets=None, predictions=None):
     if isinstance(data, list):
-        return mlflow.data.from_numpy(np.array(data), targets=np.array(targets))
+        return mlflow.data.from_numpy(
+            np.array(data), targets=np.array(targets) if targets else None
```
note: This is small but important change. In the original logic, ``targets`` param is always wrapped by numpy array when the input data is list, then ``if targets is None`` doesn't work in downstream. Concretely, when we pass a list as an input data and leave targets ``None``,  it will raise a weird error. 

This is not specific to the endpoint evaluation, however, implicitly requiring targets only for list input is confusing, so I think we should fix it in this PR.

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For MLflow >= 2.11.0, :py:func:`mlflow.evaluate()` supports evaluating a model endpoint by directly passing the MLflow Deployments endpoint URI to the ``model`` argument.
+This is particularly useful when you want to evaluate a deployed model hosted by a local `MLflow Deployments Server <../deployments/index.html>`_ or `Databricks Foundation Models API and External Models <https://docs.databricks.com/en/machine-learning/model-serving/score-foundation-models.html>`_, without implementing custom prediction logic to wrap it as an MLflow model or a python function.
```
@B-Step62 the external models docs are separate: https://docs.databricks.com/en/generative-ai/external-models/index.html. Can we link those separately?

B-Step62 replied:
Yeah but it appears the linked doc is designed for both
```
Databricks Model Serving supports Foundation Models API and external models for accessing foundation models and uses a unified OpenAI-compatible API and SDK for querying them.
```

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+
+.. hint::
+
+    When you want to use an external endpoint **not** hosted by an MLflow Deployments Server or Databricks, you can create a custom Python function following the :ref:`Evaluating with a Custom Function <llm-eval-custom-function>` guide and use it as the ``model`` argument.
```
```suggestion
    When you want to use an endpoint **not** hosted by an MLflow Deployments Server or Databricks, you can create a custom Python function following the :ref:`Evaluating with a Custom Function <llm-eval-custom-function>` guide and use it as the ``model`` argument.
```

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+
+      - For this input format, MLflow will construct the appropriate request payload to the model endpoint type. For example, if your model is a chat endpoint (``llm/v1/chat``), MLflow will wrap your input string with the chat messages format like ``{"messages": [{"role": "user", "content": "What is MLflow?"}]}``. If you want to customize the request payload e.g. including system prompt, please u ...
+
+    * - A pandas DataFrame with a dictionary column.
```
Can we bit a bit more specific in the description about what that dictionary column should contain? I don't think it should contain a `max_tokens` key, just the `messages` key for now.

B-Step62 replied:
Hmm but if customer's data is fetched from logs/traces, they should contain the params as well? I think it doesn't harm allowing that but reduces preprocessing effort.

dbczumar replied:
Fair enough! Let's keep it :)

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+            [
+                [
+                    {
+                        "messages": [
```
Can we bit a bit more specific in the description about what that dictionary column should contain - i.e. the `messages` key for the MLflow chat completions format (would also be great to link to docs for that if we have them)? I don't think it should contain a `max_tokens` key, just the `messages` key for now.

B-Step62 replied:
Will add link to https://mlflow.org/docs/latest/llms/deployments/index.html#standard-query-parameters :)

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+
+.. note::
+
+    When your input is a dictionary format tha represents request payload, it can also include the parameters like ``max_tokens``. If there are overlapping parameters in both the ``inference_params`` and the input data, the values in the ``inference_params`` will take precedence.
```
I don't think we should allow this for now (see other comments).

dbczumar replied:
```suggestion
    When your input is a dictionary format that represents request payload, it can also include the parameters like ``max_tokens``. If there are overlapping parameters in both the ``inference_params`` and the input data, the values in the ``inference_params`` will take precedence.
```

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,178 @@ up OpenAI authentication to run the code below.
+            model_type="question-answering",
+        )
+
+**Completion Endpoint hosted on** `Databricks Foundation Models API and External Models <https://docs.databricks.com/en/machine-learning/model-serving/score-foundation-models.html>`_
```
Can we separate the links for foundation models and external models (https://docs.databricks.com/en/generative-ai/external-models/index.html) throughout all the docs & docstrings?

dbczumar commented on mlflow/metrics/genai/genai_metric.py:
```diff
@@ -275,7 +275,7 @@ def eval_fn(
             raise MlflowException(
                 message="The model argument must be a string URI referring to an openai model "
-                "(openai:/gpt-3.5-turbo) or  gateway (gateway:/my-route), "
+                "(openai:/gpt-3.5-turbo) or MLflow deployment endpoint (endpoint:/my-route), "
```
```suggestion
                "(openai:/gpt-3.5-turbo) or MLflow Deployment endpoint (endpoints:/my-endpoint), "
```

dbczumar commented on mlflow/metrics/genai/model_utils.py:
```diff
@@ -113,7 +113,19 @@ def _call_openai_api(openai_uri, payload, eval_parameters):
+        deployment_uri: The URI of the deployment endpoint.
+        payload: The input payload to send to the endpoint.
+        eval_parameters: The evaluation parameters to send to the endpoint.
+        construct_payload: Whether to wrap the payload in a expected key by the endpoint,
```
Nit: This arg name is out of date. Should be `wrap_payload`.

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -425,9 +425,50 @@ def _hash_uint64_ndarray_as_bytes(array):
+    if isinstance(data, list):
+        return len(data) == 0
+    elif isinstance(data, np.ndarray):
+        return data.size == 0 or data.ndim == 0
```
```
In [1]: import numpy as np

In [2]: np.array(1).ndim
Out[2]: 0
```

In this case, the array isn't empty, but it has ndim zero. Do we need to rename this method?

B-Step62 replied:
Ahh interesting, I think we don't need the ndim check indeed.

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -425,9 +425,50 @@ def _hash_uint64_ndarray_as_bytes(array):
+        if isinstance(elm, (list, np.ndarray)):
+            result += _hash_array_of_dict_as_bytes(elm)
+        elif isinstance(elm, dict):
+            result += _hash_dict_as_bytes(elm)
```
What about all other element types? seems like we ignore them - why?

B-Step62 replied:
Great catch, this is just my silly bug....

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -425,9 +425,50 @@ def _hash_uint64_ndarray_as_bytes(array):
+    return False
+
+
+def _hash_array_of_dict_as_bytes(data):
+    # NB: If an array or list contains dictionary element, it can't be hashed with
+    # pandas.util.hash_array. Hence we need to manually hash the elements here. This is
+    # particularly for the LLM use case where the input can be a list of dictionary
+    # (chat/completion payloads), so doesn't handle more complex case like nested lists.
+    result = b""
+    for elm in data:
+        if isinstance(elm, (list, np.ndarray)):
+            result += _hash_array_of_dict_as_bytes(elm)
+        elif isinstance(elm, dict):
+            result += _hash_dict_as_bytes(elm)
```
Can we move all of these hashing methods into a separate module, ideally located in `mlflow.data`? since these are all about hashing for use with MLflow Datasets

B-Step62 replied:
Agree this should be in a separate module, but for the location, I would still put this under `evaluation` module. These methods are used for hashing various input like list, numpy, dataframe, and only used for `class EvaluationDataset` digest.

dbczumar replied:
I would prefer to put all of `EvaluationDataset` in `mlflow/data` to maintain consistency with other datasets - e.g. we don't put `TensorflowDataset` in `mlflow/tensorflow`

dbczumar replied:
Feel free to tackle this in a follow-up ticket as long as we track it :)

B-Step62 replied:
> I would prefer to put all of EvaluationDataset in mlflow/data

Makes sense.

> Feel free to tackle this in a follow-up ticket as long as we track it :)

Yeah let me do that, created JIRA in next sprint (ML-39039), but may be able to pick up if no fun surprise happens in other tasks:)

dbczumar commented on tests/evaluate/test_evaluation.py:
```diff
@@ -1567,3 +1568,257 @@ def test_evaluate_with_static_dataset_error_handling_pandas_dataset():
+            },
+        ),
+    ]
+    assert all(call in call_args_list for call in expected_calls)
+    expected_metrics_subset = {"toxicity/v1/ratio", "ari_grade_level/v1/mean"}
+    if targets:
+        expected_metrics_subset.add("exact_match/v1")
+    assert expected_metrics_subset.issubset(set(eval_result.metrics.keys()))
```
Can we also add some asserts on the contents of the per-row results / metric logs from `eval_results_table`? Specifically, I'd like to make sure that the model output column for an endpoint only contains the `content` of the first `message`, not the full `_DUMMY_COMPLETION_RESPONSE`.

B-Step62 replied:
That's a great idea!

B-Step62 committed:
feedback

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

B-Step62 committed:
fix test

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

harupy reviewed (approved):
LGTM!

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,175 @@ up OpenAI authentication to run the code below.
+      - 
+        .. code-block:: python
+
+            {
+                "inputs": [
+                    "What is MLflow?",
+                    "What is Spark?",
+                ]
+            }
```
I think we're missing `pd.DataFrame` here. This is just a dict

dbczumar commented on docs/source/llms/llm-evaluate/index.rst:
```diff
@@ -494,6 +497,175 @@ up OpenAI authentication to run the code below.
+      - 
+        .. code-block:: python
+
+            {
```
I think we're missing `pd.DataFrame` here. This is just a dict

dbczumar commented on mlflow/metrics/genai/genai_metric.py:
```diff
@@ -275,8 +275,8 @@ def eval_fn(
                 message="The model argument must be a string URI referring to an openai model "
-                "(openai:/gpt-3.5-turbo) or  gateway (gateway:/my-route), "
-                f"passed {eval_model} instead",
+                "(openai:/gpt-3.5-turbo) or an MLflow Deployment endpoint "
```
```suggestion
                "(openai:/gpt-3.5-turbo) or an MLflow Deployments endpoint "
```

dbczumar commented on mlflow/models/evaluation/base.py:
```diff
@@ -1893,7 +2025,8 @@ def pred_sample(eval_df, _builtin_metrics, _artifacts_dir):
         raise MlflowException(
             message="The model argument must be a string URI referring to an MLflow model, "
-            "an instance of `mlflow.pyfunc.PyFuncModel`, a function, or None.",
+            "an MLflow deployments endpoint, an instance of `mlflow.pyfunc.PyFuncModel`, "
```
```suggestion
            "an MLflow Deployments endpoint URI, an instance of `mlflow.pyfunc.PyFuncModel`, "
```

dbczumar reviewed (approved):
LGTM with docs nits! Thanks @B-Step62

B-Step62 committed:
doc update

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>
